File: examples/go/custody_transfer.go
package main

import (
	"context"
	"fmt"
	"log"

	"github.com/coinbase/cdp-sdk/go/openapi"
)

func CustodyTransferExample() {
	ctx := context.Background()

	client, err := createCDPClient()
	if err != nil {
		log.Fatalf("Failed to create CDP client: %v", err)
	}

	// 1. List accounts
	accountsResp, err := client.ListFoundationAccountsWithResponse(ctx, &openapi.ListFoundationAccountsParams{})
	if err != nil {
		log.Fatalf("Failed to list accounts: %v", err)
	}
	if accountsResp.JSON200 == nil {
		log.Fatalf("Failed to list accounts: %s", string(accountsResp.Body))
	}
	accounts := accountsResp.JSON200.Accounts
	fmt.Printf("Found %d accounts:\n", len(accounts))
	for _, account := range accounts {
		fmt.Printf("  %s (%s) - type: %s\n", account.AccountId, account.Name, account.Type)
	}

	if len(accounts) == 0 {
		fmt.Println("No accounts found. Create one in the CDP dashboard first.")
		return
	}

	// 2. Get balances for the first account
	account := accounts[0]
	balancesResp, err := client.ListBalancesWithResponse(ctx, account.AccountId, &openapi.ListBalancesParams{})
	if err != nil {
		log.Fatalf("Failed to list balances: %v", err)
	}
	fmt.Printf("\nBalances for %s:\n", account.Name)
	for _, balance := range balancesResp.JSON200.Balances {
		fmt.Printf("  %s: %s\n", balance.Asset, balance.Amount)
	}

	// 3. Create a quoted transfer (execute=false — no funds move)
	var source openapi.CreateTransferSource
	if err := source.FromTransfersAccount(openapi.TransfersAccount{
		AccountId: account.AccountId,
		Asset:     "usd",
	}); err != nil {
		log.Fatalf("Failed to build transfer source: %v", err)
	}

	var target openapi.TransferTarget
	if err := target.FromOnchainAddress(openapi.OnchainAddress{
		Address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
		Network: "base",
		Asset:   "usdc",
	}); err != nil {
		log.Fatalf("Failed to build transfer target: %v", err)
	}

	transferResp, err := client.CreateTransferWithResponse(ctx, &openapi.CreateTransferParams{},
		openapi.CreateTransferJSONRequestBody{
			Source:  source,
			Target:  target,
			Amount:  "10.00",
			Asset:   "usd",
			Execute: false,
		},
	)
	if err != nil {
		log.Fatalf("Failed to create transfer: %v", err)
	}
	transfer := transferResp.JSON200
	fmt.Printf("\nCreated transfer %s:\n", transfer.TransferId)
	fmt.Printf("  Status: %s\n", transfer.Status)
	fmt.Printf("  Source: %s %s\n", transfer.SourceAmount, transfer.SourceAsset)
	fmt.Printf("  Target: %s %s\n", transfer.TargetAmount, transfer.TargetAsset)
	fmt.Printf("  Expires: %s\n", transfer.ExpiresAt)

	// 4. List recent transfers
	status := openapi.TransferStatus("quoted")
	transfersResp, err := client.ListTransfersWithResponse(ctx, &openapi.ListTransfersParams{Status: &status})
	if err != nil {
		log.Fatalf("Failed to list transfers: %v", err)
	}
	fmt.Printf("\n%d quoted transfers found.\n", len(transfersResp.JSON200.Transfers))
}

File: examples/rust/examples/custody_transfer.rs
use cdp_sdk::{auth::WalletAuth, types, Client, CDP_BASE_URL};
use dotenv::dotenv;
use reqwest_middleware::ClientBuilder;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv().ok();

    let wallet_auth = WalletAuth::builder().build()?;
    let http_client = ClientBuilder::new(reqwest::Client::new())
        .with(wallet_auth)
        .build();
    let client = Client::new_with_client(CDP_BASE_URL, http_client);

    // 1. List accounts
    let accounts_response = client.list_foundation_accounts().send().await?;
    let accounts = accounts_response.into_inner();
    println!("Found {} accounts:", accounts.accounts.len());
    for account in &accounts.accounts {
        println!(
            "  {} ({}) - type: {:?}",
            *account.account_id,
            account.name.as_ref().map(|n| n.as_str()).unwrap_or(""),
            account.type_
        );
    }

    if accounts.accounts.is_empty() {
        println!("No accounts found. Create one in the CDP dashboard first.");
        return Ok(());
    }

    // 2. Get balances for the first account
    let account = &accounts.accounts[0];
    let balances_response = client
        .list_balances()
        .account_id(&account.account_id)
        .send()
        .await?;
    let balances = balances_response.into_inner();
    println!("\nBalances for {}:", account.name.as_ref().map(|n| n.as_str()).unwrap_or(""));
    for balance in &balances.balances {
        let total = balance
            .amount
            .get(balance.asset.symbol.as_str())
            .map(|d| d.available.as_str())
            .unwrap_or("0");
        println!("  {}: {}", balance.asset.symbol.as_str(), total);
    }

    // 3. Create a quoted transfer (execute: false — no funds move)
    let source = types::CreateTransferSource::TransfersAccount(types::TransfersAccount {
        account_id: account.account_id.to_string(),
        asset: "usd".parse()?,
    });

    let target = types::TransferTarget::OnchainAddress(types::OnchainAddress {
        address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913".parse()?,
        network: "base".parse()?,
        asset: "usdc".parse()?,
        destination_tag: None,
    });

    let transfer_body = types::TransferRequest::builder()
        .source(source)
        .target(target)
        .amount("10.00")
        .asset("usd".parse::<types::Asset>()?)
        .execute(false);

    let transfer_response = client
        .create_transfer()
        .body(transfer_body)
        .send()
        .await?;
    let transfer = transfer_response.into_inner();
    println!(
        "\nCreated transfer {}:",
        transfer.transfer_id.as_deref().unwrap_or("unknown")
    );
    println!("  Status: {:?}", transfer.status);
    println!(
        "  Source: {} {}",
        transfer.source_amount.as_deref().unwrap_or(""),
        transfer.source_asset.as_ref().map(|a| a.as_str()).unwrap_or("")
    );
    println!(
        "  Target: {} {}",
        transfer.target_amount.as_deref().unwrap_or(""),
        transfer.target_asset.as_ref().map(|a| a.as_str()).unwrap_or("")
    );
    println!(
        "  Expires: {}",
        transfer
            .expires_at
            .map(|d| d.to_string())
            .unwrap_or_default()
    );

    // 4. List recent transfers
    let transfers_response = client
        .list_transfers()
        .status("quoted".parse::<types::TransferStatus>()?)
        .send()
        .await?;
    let transfers = transfers_response.into_inner();
    println!("\n{} quoted transfers found.", transfers.transfers.len());

    Ok(())
}
